addons/boa/tasks.py

```diff
@@ -0,0 +1,198 @@
+
+from addons.boa import settings as boa_settings
+from addons.boa.boa_error_code import BoaErrorCode
+from framework import sentry
+from framework.celery_tasks import app as celery_app
+from osf.models import OSFUser
+from osf.utils.fields import ensure_str, ensure_bytes
+from website import settings as osf_settings
+from website.mails import send_mail, ADDONS_BOA_JOB_COMPLETE, ADDONS_BOA_JOB_FAILURE
+
+logger = logging.getLogger(__name__)
+
+
+@celery_app.task(name='addons.boa.tasks.submit_to_boa')
+def submit_to_boa(host, username, password, user_guid, project_guid, query_dataset,
+                  query_file_name, file_full_path, query_download_url, output_upload_url):
+    """
+    Download Boa query file, submit it to Boa API, wait for Boa to finish the job
+    and upload result output to OSF. Send success / failure emails notifications.
+
+    A few Notes:
+        * All the parameters must be verified by the caller.
+        * Both the ``query_download_url`` and ``output_upload_url`` must be WB URL for two reasons:
+            * It generates fewer requests between OSF and WB;
+            * It has authentication passed via the headers securely.
+        * Running asyncio in celery is tricky. Refer to the discussion below for details:
+            * https://stackoverflow.com/questions/39815771/how-to-combine-celery-with-asyncio
+    """
+    return async_to_sync(submit_to_boa_async)(host, username, password, user_guid, project_guid, query_dataset,
+                                              query_file_name, file_full_path, query_download_url, output_upload_url)
+
+
+async def submit_to_boa_async(host, username, password, user_guid, project_guid, query_dataset,
+                              query_file_name, file_full_path, query_download_url, output_upload_url):
+    """
+    Download Boa query file, submit it to Boa API, wait for Boa to finish the job
+    and upload result output to OSF. Send success / failure emails notifications.
+
+    A couple of notes:
+        * This is the async function that must be wrapped with ``async_to_sync`` by the caller
+        * See notes in ``submit_to_boa()`` for details.
+    """
+
+    logger.debug('>>>>>>>> Task begins')
+    user = await sync_to_async(OSFUser.objects.get)(guids___id=user_guid)
+    cookie_value = (await sync_to_async(user.get_or_create_cookie)()).decode()
+    project_url = f'{osf_settings.DOMAIN}{project_guid}/'
+    output_file_name = query_file_name.replace('.boa', boa_settings.OUTPUT_FILE_SUFFIX)
+
+    logger.debug(f'Downloading Boa query file: user=[{user_guid}], project=[{project_guid}], '
+                 f'file_name=[{query_file_name}], full_path=[{file_full_path}], url=[{query_download_url}] ...')
+    download_request = request.Request(query_download_url)
+    download_request.add_header('Cookie', f'{osf_settings.COOKIE_NAME}={cookie_value}')
+    try:
+        boa_query = ensure_str(request.urlopen(download_request).read())
```

Comment:
From above settings, it looks like the max size of `.boa` files we'll run is 512MB, and the content of those files is loaded into memory here. If too many approach that size, it could easily run Celery out of memory.

Is that a threshold we'll likely need, or could it be significantly lowered?

Reply:
Yes, another blocker. Looking at how and where to restrict the file size.

Reply:
fixed in https://github.com/CenterForOpenScience/osf.io/pull/10481/commits/42e45b8a12ac30f03300dec59213850f039ab93f and https://github.com/CenterForOpenScience/osf.io/pull/10481/commits/6cdedefc43d2a5958a34c42ae492cf346df59c0a